File: frontend/src/stores/metrics.ts
/**
 * P1: Metrics & Dashboard Store
 */
import { create } from 'zustand';

import { metricsApi, utmApi } from '@/lib/api';
import { getApiErrorMessage } from '@/lib/api-errors';

export interface MetricSnapshot {
  id: string;
  location_id: string;
  snapshot_date: string;
  snapshot_type: string;
  calls: number;
  directions: number;
  website_clicks: number;
  profile_views: number;
  photo_views: number;
  total_reviews: number;
  new_reviews: number;
  avg_rating: number | null;
  calls_delta?: number | null;
  directions_delta?: number | null;
  website_clicks_delta?: number | null;
  attributed_post_ids: string[];
  call_value: number;
  estimated_revenue: number;
  created_at: string;
}

export interface MetricDelta {
  current: number;
  previous: number;
  delta: number;
  percent_change: number;
}

export interface DashboardMetrics {
  calls: MetricDelta;
  directions: MetricDelta;
  website_clicks: MetricDelta;
  profile_views: MetricDelta;
  new_reviews: MetricDelta;
  avg_rating: number | null;
  estimated_revenue: number;
}

export interface Highlight {
  type: 'increase' | 'decrease' | 'milestone';
  metric: string;
  message: string;
  value: number;
  percent: number;
}

export interface TopPost {
  id: string;
  title: string;
  published_at: string | null;
  platform: string;
  estimated_impact: string;
}

export interface ChartDataPoint {
  date: string;
  calls: number;
  directions: number;
  website_clicks: number;
}

export interface DashboardData {
  location_id: string;
  period_start: string;
  period_end: string;
  metrics: DashboardMetrics;
  highlights: Highlight[];
  top_posts: TopPost[];
  chart_data: ChartDataPoint[];
}

export interface UTMLink {
  id: string;
  original_url: string;
  utm_url: string;
  utm_source: string;
  utm_medium: string;
  utm_campaign?: string | null;
  utm_content?: string | null;
  clicks: number;
  created_at: string;
}

export interface CreateUTMLinkInput {
  original_url: string;
  campaign: string;
  post_id?: string;
  utm_source?: string;
  utm_medium?: string;
}

interface MetricsState {
  dashboard: DashboardData | null;
  snapshots: MetricSnapshot[];
  utmLinks: UTMLink[];
  isLoading: boolean;
  isLoadingSnapshots: boolean;
  isLoadingUTM: boolean;
  error: string | null;
  fetchDashboard: (locationId: string, days?: number) => Promise<void>;
  fetchSnapshots: (locationId: string, days?: number) => Promise<void>;
  fetchUTMLinks: (locationId: string) => Promise<void>;
  createUTMLink: (locationId: string, data: CreateUTMLinkInput) => Promise<UTMLink>;
  deleteUTMLink: (linkId: string) => Promise<void>;
  collectMetrics: (locationId: string) => Promise<void>;
  clearError: () => void;
}

export const useMetricsStore = create<MetricsState>((set, get) => ({
  dashboard: null,
  snapshots: [],
  utmLinks: [],
  isLoading: false,
  isLoadingSnapshots: false,
  isLoadingUTM: false,
  error: null,

  fetchDashboard: async (locationId: string, days = 30) => {
    set({ isLoading: true, error: null });
    try {
      const response = await metricsApi.getDashboard(locationId, days);
      set({ dashboard: response.data, isLoading: false });
    } catch (error) {
      set({
        error: getApiErrorMessage(error, 'Failed to fetch the dashboard'),
        isLoading: false,
      });
    }
  },

  fetchSnapshots: async (locationId: string, days = 30) => {
    set({ isLoadingSnapshots: true, error: null });
    try {
      const response = await metricsApi.getSnapshots(locationId, {
        limit: days,
        snapshot_type: 'daily',
      });
      set({ snapshots: response.data.items, isLoadingSnapshots: false });
    } catch (error) {
      set({
        error: getApiErrorMessage(error, 'Failed to fetch snapshots'),
        isLoadingSnapshots: false,
      });
    }
  },

  fetchUTMLinks: async (locationId: string) => {
    set({ isLoadingUTM: true, error: null });
    try {
      const response = await utmApi.list(locationId);
      set({ utmLinks: response.data.links ?? [], isLoadingUTM: false });
    } catch (error) {
      set({
        error: getApiErrorMessage(error, 'Failed to fetch UTM links'),
        isLoadingUTM: false,
      });
    }
  },

  createUTMLink: async (locationId: string, data: CreateUTMLinkInput) => {
    const response = await utmApi.create(locationId, {
      original_url: data.original_url,
      campaign: data.campaign,
      post_id: data.post_id,
      utm_source: data.utm_source,
      utm_medium: data.utm_medium,
    });
    const newLink = response.data as UTMLink;
    set((state) => ({ utmLinks: [newLink, ...state.utmLinks] }));
    return newLink;
  },

  deleteUTMLink: async (linkId: string) => {
    await utmApi.delete(linkId);
    set((state) => ({
      utmLinks: state.utmLinks.filter((link) => link.id !== linkId),
    }));
  },

  collectMetrics: async (locationId: string) => {
    await metricsApi.collect({
      location_id: locationId,
      snapshot_date: new Date().toISOString().slice(0, 10),
      snapshot_type: 'daily',
    });
    await Promise.all([
      get().fetchDashboard(locationId),
      get().fetchSnapshots(locationId),
    ]);
  },

  clearError: () => set({ error: null }),
}));

File: app/schemas/metrics.py
"""P1: Metrics & Attribution schemas."""

from datetime import date, datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ====================
# MetricSnapshot Schemas
# ====================

class MetricSnapshotBase(BaseModel):
    """Base metrics data."""
    calls: int = 0
    directions: int = 0
    website_clicks: int = 0
    profile_views: int = 0
    photo_views: int = 0
    total_reviews: int = 0
    new_reviews: int = 0
    avg_rating: Optional[Decimal] = None


class MetricSnapshotCreate(MetricSnapshotBase):
    """Create snapshot manually."""
    location_id: UUID
    snapshot_date: date
    snapshot_type: str = "daily"  # daily, weekly, monthly
    call_value: Decimal = Decimal("50.00")


class MetricSnapshotResponse(MetricSnapshotBase):
    """Snapshot response."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    location_id: UUID
    snapshot_date: date
    snapshot_type: str
    calls_delta: Optional[int] = None
    directions_delta: Optional[int] = None
    website_clicks_delta: Optional[int] = None
    attributed_post_ids: list[UUID] = []
    call_value: Decimal
    estimated_revenue: Decimal
    created_at: datetime


class MetricSnapshotList(BaseModel):
    """List of snapshots."""
    items: list[MetricSnapshotResponse]
    total: int


# ====================
# Dashboard Schemas
# ====================

class MetricDelta(BaseModel):
    """Metric change between periods."""
    current: int
    previous: int
    delta: int
    percent_change: float


class DashboardMetrics(BaseModel):
    """Current period metrics summary."""
    calls: MetricDelta
    directions: MetricDelta
    website_clicks: MetricDelta
    profile_views: MetricDelta
    new_reviews: MetricDelta
    avg_rating: Optional[Decimal] = None
    estimated_revenue: Decimal


class DashboardHighlight(BaseModel):
    """Dashboard highlight item."""
    type: str  # "increase", "decrease", "milestone"
    metric: str
    message: str
    value: int
    percent: float


class TopPost(BaseModel):
    """Top contributing post."""
    id: UUID
    title: str
    published_at: Optional[datetime]
    platform: str
    estimated_impact: str  # "high", "medium", "low"


class ChartDataPoint(BaseModel):
    """Single point for chart."""
    date: date
    calls: int
    directions: int
    website_clicks: int


class DashboardResponse(BaseModel):
    """Full dashboard data."""
    location_id: UUID
    period_start: date
    period_end: date
    metrics: DashboardMetrics
    highlights: list[DashboardHighlight]
    top_posts: list[TopPost]
    chart_data: list[ChartDataPoint]


# ====================
# Weekly Report Schemas
# ====================

class WeeklyReportSummary(BaseModel):
    """Report summary data."""
    calls_total: int
    calls_delta: int
    calls_percent: float
    directions_total: int
    directions_delta: int
    directions_percent: float
    website_clicks_total: int
    new_reviews: int
    avg_rating: Optional[Decimal]
    estimated_revenue: Decimal
    top_day: str
    highlights: list[str]


class WeeklyReportCreate(BaseModel):
    """Create weekly report."""
    location_id: UUID
    report_week: date  # Monday of the week


class WeeklyReportResponse(BaseModel):
    """Weekly report response."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    location_id: UUID
    report_week: date
    report_type: str
    summary: WeeklyReportSummary
    pdf_url: Optional[str] = None
    sent_at: Optional[datetime] = None
    sent_to: list[str] = []
    created_at: datetime


class WeeklyReportList(BaseModel):
    """List of reports."""
    items: list[WeeklyReportResponse]
    total: int


class SendReportRequest(BaseModel):
    """Send report via email."""
    email_addresses: list[str]


# ====================
# UTM Link Schemas
# ====================

class UTMGenerateRequest(BaseModel):
    """Generate UTM link."""
    original_url: str
    campaign: Optional[str] = None
    post_id: Optional[UUID] = None
    utm_source: Optional[str] = None
    utm_medium: Optional[str] = None


class UTMLinkResponse(BaseModel):
    """UTM link response."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    original_url: str
    utm_url: str
    utm_source: str
    utm_medium: str
    utm_campaign: Optional[str]
    utm_content: Optional[str]
    clicks: int
    created_at: datetime


class UTMStatsResponse(BaseModel):
    """UTM stats."""
    total_links: int
    total_clicks: int
    links: list[UTMLinkResponse]
